File: dsa/topics/00_basics/java/BubbleSort.java
import java.util.Arrays;

public class BubbleSort {

	public static void main(String[] args) {
		BubbleSort bsort = new BubbleSort();
		int[] numbers = {2,7,4,1,5,3};
		System.out.println(Arrays.toString(numbers));
		bsort.bubbleSort(numbers);

	}

	// Idea: The biggest element is bubbled to last one for each iteration
	// Time complexity in worst & average case is O(n^2). Best case O(n)
	// Stable sorting
	// Inplace sorting.
	public void bubbleSort(int[] numbers) {
		for(int i=0; i<numbers.length; i++){
			boolean sorted = true;
			for(int j=0; j<numbers.length-1; j++){
				if(numbers[j] > numbers[j+1]){
					int temp = numbers[j];
					numbers[j] = numbers[j+1];
					numbers[j+1] = temp;

					sorted = false;
				}
			}
			if(sorted){
				break;
			}
			System.out.println(Arrays.toString(numbers));
		}

	}

}

File: dsa/topics/00_basics/python/bubble_sort.py
"""
Bubble Sort

Idea: The biggest element is bubbled to last one for each iteration
Time complexity in worst & average case is O(n^2). Best case O(n)
Stable sorting
Inplace sorting.
"""

def bubble_sort(nums):
    n = len(nums)
    
    for i in range(n):
        sorted_flag = True
        
        for j in range(n - 1):
            if nums[j] > nums[j + 1]:
                nums[j], nums[j + 1] = nums[j + 1], nums[j]
                sorted_flag = False
        
        if sorted_flag:
            break
        
        print(nums)


if __name__ == "__main__":
    numbers = [2, 7, 4, 1, 5, 3]
    print(f"Original: {numbers}")
    bubble_sort(numbers)
    print(f"Final sorted: {numbers}")
